### PR TITLE
Validate Punycoded TLDs

### PR DESCRIPTION
**Pull request by BreyndotEchse**
Method ``isValid`` must return true for IDN top level domains.

### CODE REVIEW

File: test/HostnameTest.php [hunk 1]
```diff
@@ -506,6 +506,25 @@ public function testAdditionalUTF8TLDs()
         }
     }
 
+    public function testAdditionalPunycodedTLDs()
+    {
+        $validator = new Hostname(Hostname::ALLOW_ALL);
+        // Check UTF-8 TLD matching
+        $valuesExpected = [
```
Comment (Maks3w) [inline test/HostnameTest.php]: Please refactor using data providers.

Also split the test in two tests, one when is valid and another one when is not

Comment (Maks3w) [inline test/HostnameTest.php]: When refactor please add a description for each case using the key of the array

`['invalid foo' => 'xn--3-owe4au9mpa.xn--xkc2al3hye2a']`

File: src/Hostname.php [hunk 2]
```diff
@@ -1772,8 +1772,16 @@ public function isValid($value)
                     // id-prefix: alpha / digit
                     // ldh: alpha / digit / dash
 
-                    // Match TLD against known list
                     $this->tld = strtoupper($matches[1]);
+                    // Decode Punycode TLD to IDN
+                    if (strpos($this->tld, 'xn--') === 0) {
+                        $this->tld = $this->decodePunycode(substr($this->tld, 4));
```
Comment (Maks3w) [inline src/Hostname.php]: Setting multibyte values in the property breaks non mutlibyte methods `if (!in_array(strtolower($this->tld), $this->validTlds)`

Comment (Maks3w) [inline src/Hostname.php]: TLD attribute must be ASCII for keep the backward compatibility

File: test/HostnameTest.php (hunk 1 above)
Comment (BreyndotEchse) [inline test/HostnameTest.php]: Only for ``testAdditionalPunycodedTLDs`` or also for ``testAdditionalUTF8TLDs`` (merged)?

Example dataProvider (valid hostnames):
```
public function validTLDHostnames()
{
    // @codingStandardsIgnoreStart
    return [
        'test123.онлайн' => 'test123.онлайн',
        'test123.онлайн' => 'test123.xn--80asehdb',
        'тест.рф'        => 'тест.рф',
        'тест.рф'        => 'xn--e1aybc.xn--p1ai',
        'туршилтын.мон'  => 'туршилтын.мон',
        'туршилтын.мон'  => 'xn--h1aggjjdd5b4a.xn--l1acc',
    ];
    // @codingStandardsIgnoreEnd
}
```

Comment (Maks3w) [inline test/HostnameTest.php]: I'll appreciate if you can do it testAdditionalUTF8TLDs as well.

What I miss with the descriptions I don't quickly understand why that case is significative.

```
return [
        'UTF8 TLD' => 'test123.онлайн',
        'Punyencoded TLD' => 'test123.xn--80asehdb',
...
    ];
```

File: src/Hostname.php (hunk 2 above)
Comment (BreyndotEchse) [inline src/Hostname.php]: But the current behaviour is to keep the UTF-8 TLD. For ``$hostnameValidator->isValid('тест.рф')`` the property ``$tld`` is ``рф``. Tested with latest master.

``Hostname::$validIdns`` uses UTF-8 TLDs as keys and ``Hostname::$validTlds`` holds the UTF-8 TLDs.

**Pull request merged**